File: 2398-maximum-number-of-robots-within-budget/2398-maximum-number-of-robots-within-budget.py
class Solution:
    def maximumRobots(self, chargeTimes: List[int], runningCosts: List[int], budget: int) -> int:
        dq = deque()
        n = len(chargeTimes)
        j = s = 0
        for i, cost in enumerate(runningCosts):
            s += cost
            while dq and chargeTimes[dq[-1]] <= chargeTimes[i]:
                dq.pop()
            dq.append(i)
            if dq and chargeTimes[dq[0]] + s * (i - j + 1) > budget:
                if dq[0] == j:
                    dq.popleft()
                s -= runningCosts[j]
                j += 1
        return n - j

File: 2398-maximum-number-of-robots-within-budget/2398-maximum-number-of-robots-within-budget.js
/**
 * @param {number[]} chargeTimes
 * @param {number[]} runningCosts
 * @param {number} budget
 * @return {number}
 */
var maximumRobots = function(chargeTimes, runningCosts, budget) {
  const queue = [];
  const n = chargeTimes.length;
  let [l, r, start, sum] = [0, -1, 0, 0];
  
  for (let i = 0; i < n; i++) {
    sum += runningCosts[i];
    while (l <= r && chargeTimes[queue[r]] <= chargeTimes[i]) r--;
    
    queue[++r] = i;
    
    if (l <= r && chargeTimes[queue[l]] + (i - start + 1) * sum > budget) {
      if (queue[l] === start) l++;
      
      sum -= runningCosts[start++];
    }
  }
  
  return n - start;
};